src/crab/type_domain.hpp

```diff
@@ -0,0 +1,357 @@
+
+        return ((hash<int>()(reg.r)
+               ^ (hash<int>()(reg.loc.first.from) << 1)) >> 1)
+               ^ (hash<int>()(reg.loc.second) << 1);
+    }
+};
+
+namespace crab {
+
+
+using offset_to_ptr_no_off_t = std::unordered_map<uint64_t, ptr_no_off_t>;
+using offset_to_ptr_t = std::unordered_map<uint64_t, ptr_t>;
+
+struct ctx_t {
+  private:
+    offset_to_ptr_no_off_t packet_ptrs;
+
+  public:
+    ctx_t(const ebpf_context_descriptor_t* desc)
+    {
+        packet_ptrs.insert(std::make_pair(desc->data, crab::ptr_no_off_t(crab::region::T_PACKET)));
+        packet_ptrs.insert(std::make_pair(desc->end, crab::ptr_no_off_t(crab::region::T_PACKET)));
+    }
+
+    std::optional<ptr_no_off_t> find(int key) const {
+        auto it = packet_ptrs.find(key);
+        if (it == packet_ptrs.end()) return {};
+        return it->second;
+    }
+};
+
+struct stack_t {
+  private:
+    offset_to_ptr_t ptrs;
+    bool _is_bottom;
+
+  public:
+    explicit stack_t(bool is_bottom = false) : _is_bottom(is_bottom) {}
+
+    stack_t operator|(const stack_t& other) const {
+        stack_t st{};
+        for (auto& e : ptrs) {
+            auto it = other.ptrs.find(e.first);
+            if (it == other.ptrs.end() || it->second == e.second) {
+                st.ptrs.insert(e);
+            }
+        }
+
+        for (auto& e : other.ptrs) {
+            auto it = ptrs.find(e.first);
+            if (it == ptrs.end()) {
+                st.ptrs.insert(e);
+            }
+        }
+        return st;
+    }
+
+    void set_to_bottom() {
+        this->~stack_t();
+        new (this) stack_t(true);
+    }
+
+    void set_to_top() {
+        this->~stack_t();
+        new (this) stack_t(false);
+    }
+
+    static stack_t bottom() { return stack_t(true); }
+
+    static stack_t top() { return stack_t(false); }
+
+    bool is_bottom() const { return _is_bottom; }
+
+    bool is_top() const {
+        if (_is_bottom)
+            return false;
+        return ptrs.empty();
+    }
+
+    offset_to_ptr_t get_ptrs() const { return ptrs; }
+
+    void insert(int key, ptr_t value) { ptrs.insert(std::make_pair(key, value)); }
+
+    std::optional<ptr_t> find(int key) {
+        auto it = ptrs.find(key);
+        if (it == ptrs.end()) return {};
+        return it->second;
+    }
+
+    friend std::ostream& operator<<(std::ostream& o, const stack_t& st) {
+        o << "Stack: {";
+        for (auto s : st.get_ptrs()) {
+            o << s.first << ": ";
+            if (std::holds_alternative<ptr_with_off_t>(s.second)) {
+                auto t = std::get<ptr_with_off_t>(s.second);
+                o << t << ", ";
+            }
+            else {
+                auto t = std::get<ptr_no_off_t>(s.second);
+                o << t << ", ";
+            }
+        }
+        return o << "}";
+    }
+};
+
+using all_types_t = std::unordered_map<reg_with_loc_t, ptr_t>;
+using reg_live_vars_t = std::array<reg_with_loc_t, 11>;
+
+struct types_t {
```

Comment:
The name `types_t` is not very helpful. This class is supposed to keep the types for registers. So please call it something like `register_types_t`